src/main/java/io/confluent/connect/elasticsearch/ElasticsearchClient.java

```diff
@@ -15,102 +15,568 @@
+                config.maxRetries()
+            )
+        )
+        .build();
+  }
+
+  /**
+   * Returns the underlying Elasticsearch client.
+   *
+   * @return the underlying RestHighLevelClient
+   */
+  public RestHighLevelClient client() {
+    return client;
+  }
+
+  /**
+   * Closes the ElasticsearchClient.
+   *
+   * @throws ConnectException if all of the records fail to flush before the timeout.
+   */
+  public void close() {
+    try {
+      if (!bulkProcessor.awaitClose(1, TimeUnit.MINUTES)) {
+        throw new ConnectException("Failed to process all outstanding requests in time.");
+      }
+    } catch (InterruptedException e) {
+      log.error("Bulk processor close was interrupted.", e);
+    }
+  }
+
+  /**
+   * Creates an index. Will not recreate the index if it already exists.
+   *
+   * @param index the index to create
+   */
+  public void createIndex(String index) {
+    if (indexExists(index)) {
+      return;
+    }
+
+    CreateIndexRequest request = new CreateIndexRequest(index);
+    callWithRetries(
+        "create index " + index,
+        () -> {
+          try {
+            client.indices().create(request, RequestOptions.DEFAULT);
+          } catch (ElasticsearchStatusException | IOException e) {
+            if (!e.getMessage().contains(RESOURCE_ALREADY_EXISTS_EXCEPTION)) {
+              throw e;
+            }
+          }
+
+          return null;
+        }
+    );
+  }
+
+  /**
+   * Creates a mapping for the given index and schema.
+   *
+   * @param index the index to create the mapping for
+   * @param schema the schema to map
+   */
+  public void createMapping(String index, Schema schema) {
+    PutMappingRequest request = new PutMappingRequest(index).source(Mapping.buildMapping(schema));
+    callWithRetries(
+        "create mapping",
+        () -> client.indices().putMapping(request, RequestOptions.DEFAULT)
+    );
+  }
+
+  /**
+   * Flushes any buffered records.
+   */
+  public void flush() {
+    bulkProcessor.flush();
+  }
+
+  /**
+   * Checks whether the index already has a mapping or not.
+   * @param index the index to check
+   * @return true if a mapping exists, false if it does not
+   */
+  public boolean hasMapping(String index) {
+    MappingMetaData mapping = getMapping(index);
+    return mapping != null && !mapping.sourceAsMap().isEmpty();
+  }
+
+  /**
+   * Buffers a record to index. Will ensure that there are no concurrent requests for the same
+   * document id.
+   *
+   * @param record the record to index
+   * @param request the associated request to send
+   * @throws ConnectException if one of the requests failed
+   */
+  public void index(SinkRecord record, DocWriteRequest<?> request) {
+    if (isFailed()) {
+      close();
+      throw error.get();
+    }
+
+    if (recordMap != null) {
+      // make sure that only unique records are being sent in a batch
+      while (recordMap.contains(request.id())) {
```

Comment:
Seems you should be looking for the key, right?. `contains` looks for the value. It's also a legacy method. Using `ConcurrentMap` will help you avoid this issue.